### PR TITLE
Simplify tests that should cover all possible devices

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
This PR introduce `@pytorchtest.for_all_device_types()` decorator which helps to write CPU, CUDA tests faster, iterating single test through all available devices

Simple `test_var_mean_some_dims`  becomes
```
test_var_mean_some_dims (__main__.TestTorch) ... ok
test_var_mean_some_dims_cpu (__main__.TestTorch) ... ok
test_var_mean_some_dims_cuda (__main__.TestTorch) ... ok
```

```python

class pytorchtest():
    """Allows to generate and run per-device unittests.

    This decorator class allows to generate and run per-device unittest.

    Example:

    class _TestTorchMixin(pytorchtest):

        @pytorchtest.for_all_device_types()
        def test_zeros_like(self, device):
            expected = torch.zeros((100, 100,), device=device)

    Will execute:

        test_zeros_like (__main__.TestTorch) ... skipped 'Look at test_zeros_like_cpu, test_zeros_like_cuda results.'
        test_zeros_like_cpu (__main__.TestTorch) ... ok
        test_zeros_like_cuda (__main__.TestTorch) ... ok

    To work properly, test class should be inherited from the `pytorchtest`.
    for_all_device_types decorator does not guarantee proper functionality in
    combination with other decorators.

    Please do not extend this decorator to support other cases (such as dtype,
    layouts, etc) without consulting with bigger group. Devices is the special
    case as build flags control additions/removals (see
    https://github.com/pytorch/pytorch/pull/23824 for the reference).
    """
```

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
cc @ezyang @gchanan recommended review mode 'hide whitespaces changes'. So far it is just a proposal. If you agree with such implementation I will add better docs and convert more tests ( ~ 50 found only inside of the test_torch.py )

**Comment by xuhdev** [comment]
@VitalyFedyunin Is it possible to run only CPU test or CUDA test? Sometimes, even if I have a CUDA build, I still would like to avoid running the full CUDA test because they often just run out of memory.

**Comment by VitalyFedyunin** [comment]
> @VitalyFedyunin Is it possible to run only CPU test or CUDA test? Sometimes, even if I have a CUDA build, I still would like to avoid running the full CUDA test because they often just run out of memory.

Yes, test names propagation happens before filtering, so you can always do:

```python test/test_torch.py --verbose -k test_var_mean_some_dims_cpu
test_var_mean_some_dims_cpu (__main__.TestTorch) ... ok
```

**Comment by gchanan** [comment]
can you describe the motivation?  Having to know the "_cpu" magic is IMO pretty bad, and once you add one of these, more will pile on, and we'll be remembering arcane knowledge like the correct order of "mytest_cpu_half_contiguous" or something ridiculous.

**Comment by gchanan** [comment]
put another way: I almost never want to actually run one of these at a time, and if I do, it's trivial to modify the test to run it.

**Comment by VitalyFedyunin** [comment]
Motivation is to promote one method over everything else, as currently we have:

1) `use_cuda` variants
```python

    def _test_random_neg_values(self, use_cuda=False):
        signed_types = ['torch.DoubleTensor', 'torch.FloatTensor', 'torch.LongTensor',
                        'torch.IntTensor', 'torch.ShortTensor']
        for tname in signed_types:
            res = torch.rand(SIZE, SIZE).type(tname)
            if use_cuda:
                res = res.cuda()
            res.random_(-10, -1)
            self.assertLessEqual(res.max().item(), 9)
            self.assertGreaterEqual(res.min().item(), -10)

    def test_random_neg_values(self):
        self._test_random_neg_values(self)
```

Requires 3 function declarations and weird `if` statements.

2) `get_all_device_types` variants:

```python
    def test_add(self):
        for device in torch.testing.get_all_device_types():
            # [res] torch.add([res,] tensor1, tensor2)
            m1 = torch.randn(100, 100, device=device)
            v1 = torch.randn(100, device=device)
# ......
```
Generally good, but when it fails, you spent time to figure out if it is CPU error or CUDA error.
Also it is harder to filter tests by name.

1) _cuda and _cpu variants 

```python

    def test_all_any_empty(self):
        x = torch.ByteTensor()
# .....

    @unittest.skipIf(not torch.cuda.is_available(), 'no CUDA')
    def test_all_any_empty_cuda(self): 
# .......
```

Lots of code duplication, prone to error.

**Comment by gchanan** [comment]
Well I agree `use_cuda` variants and `_cuda` and `_cpu` are bad :).

I'm not sure why from the errors from the for loop case are hard to determine -- can you give an example?

**Comment by SsnL** [comment]
> I'm not sure why from the errors from the for loop case are hard to determine -- can you give an example?

For me they are hard because I can't tell which loop it is falling at. E.g., if the loop is through [`'cpu', 'cuda']`, I can't tell from the trace if the `cpu` or the `cuda` one is broken.

**Comment by ezyang** [comment]
In general, here are the (possibly mutually exclusive) constraints I'd like to see in an ideal scenario:

1. When a test `test_var_mean_some_dims_cpu` fails, it is obvious how to find the definition of this test. (Implicit addition of suffixes makes it harder to find tests: if you grep for the test name, you find that the string doesn't occur anywhere in the codebase.)
2. When a test that runs multiple tests fails, it is easy to tell what exactly failed without having to modify the test (@SsnL's comment)
3. It is easy to run a test only under one configuration. (@xuhdev's comment) Having to edit the test manually is more annoying than being able to just run a particular test name to get only a single configuration
4. Make it difficult to impossible to forget to run the test under other configurations. Part of this is minimizing the boilerplate necessary to test under multiple configurations.

Sometimes, violating some of these constraints just causes mild annoyance (e.g., @gchanan thinks it is not a big deal to edit a test case to do only one configuration (3)--I tend to agree); in other cases, we should try very hard not to violate the constraint, because we are trying to solve some systemic problem in the test suite (e.g., as per (4)).

I personally like what we did in `test/test_sparse.py`, where we defined factory functions on the test class, and then subclassed the test class to test under other configurations (e.g., CUDA) because it involves less boilerplate than having to stick a decorator on every method. But @VitalyFedyunin, if you're doing the work, I think you should make the final call :)

**Comment by VitalyFedyunin** [comment]
I like `test/test_sparse.py` solution, but it is not applicable in the `test_torch.py` as there are tests which are cross-device. 

Decorator that I'm proposing covers 2,3,4 from the @ezyang  list and partially covers 1 (you need to remember to grep everything before postfix). I purposely didn't modify call stack, so it is possible to find function name in it:
```
======================================================================
FAIL: test_var_mean_some_dims_cpu (__main__.TestTorch)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/vitalyf/pytorch_wp2/test/common_utils.py", line 213, in run_test_function
    return fn(self, device)
  File "test/test_torch.py", line 2811, in test_var_mean_some_dims
    self.assertEqual('should_fail', device)
  File "/data/users/vitalyf/pytorch_wp2/test/common_utils.py", line 647, in assertEqual
    super(TestCase, self).assertEqual(x, y, message)
AssertionError: 'should_fail' != 'cpu'
```

**Comment by gchanan** [comment]
I actually am not a fan of test_sparse :).  You have to remember to use non-standard constructors (`ValueTensor`!) or you keep the name of the constructors but give them non-standard definitions.

I spoke a little bit with @VitalyFedyunin about this.  For finding which test was run, he gave the example of:
`assertTrue(my_op(my_input).has_memory_format(some_memory_format))`

now that may fail only on cuda but you just see the memory format doesn't match.  My personal opinion on that is the assert string should be set so you can see that, but I do think reasonable minds can differ on that, because it's a little unlikely we'd be able to maintain that.

So, I think either of the two approaches are acceptable:
1) write out for loops, make sure assertions give nice messages
2) have _only_ a device decorator, comment that it's special because e.g. we support build flags with additions/removals of devices and that new decorators cannot be added without approvals and even then, they are NOT composable (to avoid `test_dims_cuda_noncontiguous_float`).

**Comment by gchanan** [comment]
one other question: can I run a single command to run a test across all devices?  Or do I need to use pytest for that?

**Comment by VitalyFedyunin** [comment]
> one other question: can I run a single command to run a test across all devices? Or do I need to use pytest for that?

Decorator does not change the way we start tests, you can run test for all devices with simple
`python test/test_torch.py --verbose`

**Comment by gchanan** [comment]
I meant a single test, e.g. `test_var_mean_some_dims_cpu` and `test_var_mean_some_dims_cuda` in one command.

**Comment by VitalyFedyunin** [comment]
`python test/test_torch.py --verbose -k test_memory_format_resize_as_cuda` will run single for CUDA

File: test/test_torch.py
Comment (gchanan) [inline test/test_torch.py]: nit: can we call this `torchtest`?  We don't actually use the name "PyTorch" much in the code...

File: test/common_utils.py
Comment (gchanan) [inline test/common_utils.py]: nit: `(fn, device)`

File: test/common_utils.py
Comment (gchanan) [inline test/common_utils.py]: nit: you don't need `the` in this sentence.

File: test/common_utils.py
Comment (gchanan) [inline test/common_utils.py]: you don't need to change this, just noting that it's a little surprising the device is a string here and not a device.  Do you think that should be changed before we roll this out?

**Review by gchanan — DISMISSED** [review]

**Comment by ezyang** [comment]
> I actually am not a fan of test_sparse :). You have to remember to use non-standard constructors (ValueTensor!) or you keep the name of the constructors but give them non-standard definitions.

Don't consider blocking for the PR, but I do want to point out one defense here: you don't have to do non-standard constructors, you could always just have the current dtype be a member on the class!

File: test/common_utils.py
Comment (VitalyFedyunin) [reply test/common_utils.py]:
```python
def get_all_device_types():
    return ['cpu'] if not torch.cuda.is_available() else ['cpu', 'cuda']
```

**Comment by VitalyFedyunin** [comment]
Note for myself:

Found one more pattern

```python
def _test_dim_reduction(self, cast):
    # bla bla bla
    return

def test_dim_reduction(self):
    self._test_dim_reduction(self, lambda t: t)
```

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**
